File: kotlin/src/test/kotlin/org/meshtastic/tak/CotTypeMapperTest.kt
package org.meshtastic.tak

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*

class CotTypeMapperTest {

    @Test
    fun `known type strings map to correct enum values`() {
        assertEquals(CotTypeMapper.COTTYPE_A_F_G_U_C, CotTypeMapper.typeToEnum("a-f-G-U-C"))
        assertEquals(CotTypeMapper.COTTYPE_A_N_A_C_F, CotTypeMapper.typeToEnum("a-n-A-C-F"))
        assertEquals(CotTypeMapper.COTTYPE_T_X_D_D, CotTypeMapper.typeToEnum("t-x-d-d"))
        assertEquals(CotTypeMapper.COTTYPE_B_T_F, CotTypeMapper.typeToEnum("b-t-f"))
        assertEquals(CotTypeMapper.COTTYPE_B_R_F_H_C, CotTypeMapper.typeToEnum("b-r-f-h-c"))
        assertEquals(CotTypeMapper.COTTYPE_B_A_O_OPN, CotTypeMapper.typeToEnum("b-a-o-opn"))
        assertEquals(CotTypeMapper.COTTYPE_U_D_F, CotTypeMapper.typeToEnum("u-d-f"))
    }

    @Test
    fun `unknown type strings map to OTHER`() {
        assertEquals(CotTypeMapper.COTTYPE_OTHER, CotTypeMapper.typeToEnum("a-f-G-U-C-X-Y-Z"))
        assertEquals(CotTypeMapper.COTTYPE_OTHER, CotTypeMapper.typeToEnum("z-unknown"))
        assertEquals(CotTypeMapper.COTTYPE_OTHER, CotTypeMapper.typeToEnum(""))
    }

    @Test
    fun `enum values round-trip through string`() {
        for (enumVal in 1..75) {
            val str = CotTypeMapper.typeToString(enumVal)
            if (str != null) {
                assertEquals(enumVal, CotTypeMapper.typeToEnum(str),
                    "Round-trip failed for enum $enumVal -> $str")
            }
        }
    }

    @Test
    fun `OTHER returns null string`() {
        assertNull(CotTypeMapper.typeToString(CotTypeMapper.COTTYPE_OTHER))
    }

    @Test
    fun `aircraft classification is correct`() {
        // Air domain types
        assertTrue(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_N_A_C_F))
        assertTrue(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_F_A_M_H))
        assertTrue(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_H_A_M_F_F))
        assertTrue(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_U_A_C))
        assertTrue(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_H_A))
        assertTrue(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_U_A))

        // Non-air domain types
        assertFalse(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_F_G_U_C))
        assertFalse(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_T_X_D_D))
        assertFalse(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_B_T_F))
        assertFalse(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_U_D_F))
        assertFalse(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_F_G))
        assertFalse(CotTypeMapper.isAircraft(CotTypeMapper.COTTYPE_A_F_S))
    }

    @Test
    fun `aircraft string classification is correct`() {
        assertTrue(CotTypeMapper.isAircraftString("a-f-A-M-H"))
        assertTrue(CotTypeMapper.isAircraftString("a-n-A-C-F"))
        assertTrue(CotTypeMapper.isAircraftString("a-h-A-M-F-F"))

        assertFalse(CotTypeMapper.isAircraftString("a-f-G-U-C"))
        assertFalse(CotTypeMapper.isAircraftString("b-t-f"))
        assertFalse(CotTypeMapper.isAircraftString("t-x-d-d"))
        assertFalse(CotTypeMapper.isAircraftString("a-f-S"))
    }

    @Test
    fun `how strings map correctly`() {
        assertEquals(CotTypeMapper.COTHOW_H_E, CotTypeMapper.howToEnum("h-e"))
        assertEquals(CotTypeMapper.COTHOW_M_G, CotTypeMapper.howToEnum("m-g"))
        assertEquals(CotTypeMapper.COTHOW_H_G_I_G_O, CotTypeMapper.howToEnum("h-g-i-g-o"))
        assertEquals(CotTypeMapper.COTHOW_M_R, CotTypeMapper.howToEnum("m-r"))
        assertEquals(CotTypeMapper.COTHOW_UNSPECIFIED, CotTypeMapper.howToEnum("unknown-how"))
    }

    @Test
    fun `how enum round-trips through string`() {
        for (enumVal in 1..7) {
            val str = CotTypeMapper.howToString(enumVal)
            if (str != null) {
                assertEquals(enumVal, CotTypeMapper.howToEnum(str),
                    "How round-trip failed for enum $enumVal -> $str")
            }
        }
    }
}

File: typescript/tests/cotTypeMapper.test.ts
import { describe, it, expect } from "vitest";
import { typeToEnum, typeToString, howToEnum, howToString, isAircraft, isAircraftString, COTTYPE_OTHER } from "../src/CotTypeMapper.js";

describe("CotTypeMapper", () => {
  it("maps known types correctly", () => {
    expect(typeToEnum("a-f-G-U-C")).toBe(1);
    expect(typeToEnum("a-n-A-C-F")).toBe(3);
    expect(typeToEnum("t-x-d-d")).toBe(14);
    expect(typeToEnum("b-t-f")).toBe(25);
    expect(typeToEnum("b-r-f-h-c")).toBe(26);
  });

  it("returns OTHER for unknown types", () => {
    expect(typeToEnum("z-unknown")).toBe(0);
    expect(typeToEnum("")).toBe(0);
  });

  it("OTHER returns undefined string", () => {
    expect(typeToString(COTTYPE_OTHER)).toBeUndefined();
  });

  it("classifies aircraft strings correctly", () => {
    expect(isAircraftString("a-f-A-M-H")).toBe(true);
    expect(isAircraftString("a-n-A-C-F")).toBe(true);
    expect(isAircraftString("a-f-G-U-C")).toBe(false);
    expect(isAircraftString("b-t-f")).toBe(false);
    expect(isAircraftString("a-f-S")).toBe(false);
  });

  it("classifies aircraft string edge cases correctly", () => {
    // Fewer than 3 atoms -> always non-aircraft (guards atoms.length >= 3)
    expect(isAircraftString("")).toBe(false);
    expect(isAircraftString("A")).toBe(false);
    expect(isAircraftString("a")).toBe(false);
    expect(isAircraftString("a-f")).toBe(false);

    // 'A' present but NOT at index 2 -> non-aircraft
    expect(isAircraftString("A-f-G")).toBe(false); // index 0
    expect(isAircraftString("a-A-G")).toBe(false); // index 1
    expect(isAircraftString("a---A")).toBe(false); // index 3 after empty atoms
    expect(isAircraftString("a-f-G-A")).toBe(false); // index 3

    // Exactly 3 atoms with A at index 2 -> minimal valid aircraft type
    expect(isAircraftString("a-f-A")).toBe(true);
    expect(isAircraftString("a-h-A")).toBe(true);
  });

  it("round-trips type enum <-> string", () => {
    for (let i = 1; i <= 75; i++) {
      const s = typeToString(i);
      if (s) expect(typeToEnum(s)).toBe(i);
    }
  });

  it("classifies aircraft correctly", () => {
    expect(isAircraft(3)).toBe(true);   // a-n-A-C-F
    expect(isAircraft(6)).toBe(true);   // a-f-A-M-H
    expect(isAircraft(11)).toBe(true);  // a-h-A-M-F-F
    expect(isAircraft(44)).toBe(true);  // a-h-A

    expect(isAircraft(1)).toBe(false);  // a-f-G-U-C
    expect(isAircraft(14)).toBe(false); // t-x-d-d
    expect(isAircraft(25)).toBe(false); // b-t-f
    expect(isAircraft(17)).toBe(false); // a-f-S
  });

  it("maps how values correctly", () => {
    expect(howToEnum("h-e")).toBe(1);
    expect(howToEnum("m-g")).toBe(2);
    expect(howToEnum("m-r")).toBe(4);
    expect(howToEnum("unknown")).toBe(0);
  });

  it("round-trips how enum <-> string", () => {
    for (let i = 1; i <= 7; i++) {
      const s = howToString(i);
      if (s) expect(howToEnum(s)).toBe(i);
    }
  });
});
